theano/sandbox/cuda/opt.py

```diff
@@ -441,6 +441,24 @@ def local_gpu_dot_to_dot22(node):
     return False
 
 
+@local_optimizer([gpu_from_host, host_from_gpu])
+def local_assert_no_cpu_op(node):
+    if not isinstance(node.op, GpuOp) and all([var.owner and isinstance(var.owner.op,
+        HostFromGpu) for var in node.inputs]) and all([var.owner and
+            isinstance(var.owner.op, GpuFromHost) for var in node.outputs]):
+            if config.assert_no_cpu_op == "warn":
+                _logger.warning(("CPU op %s is detected in the computational"
+                                 " graph") % node)
+            elif config.assert_no_cpu_op == "raise":
+                raise RuntimeError("The op %s is on CPU." % node)
+            elif config.assert_no_cpu_op == "pdb":
+                import ipdb; ipdb.set_trace()
```

Comment:
I think we usually use pdb, not ipdb, and import it at the beginning of the file.
Also, some places use `pdb.post_mortem(sys.exc_info()[2])` instead of `pdb.set_trace()`, but I'm not sure what difference it makes.